web/src/components/graph/areagraph.js

```diff
@@ -0,0 +1,141 @@
+import React, {
+  useState,
+  useEffect,
+  useRef,
+  useMemo,
+} from 'react';
+import {
+  first,
+  last,
+  max,
+  isNumber,
+} from 'lodash';
+import { scaleTime, scaleLinear } from 'd3-scale';
+import moment from 'moment';
+
+import AreaGraphLayers from './areagraphlayers';
+import GraphBackground from './graphbackground';
+import GraphHoverLine from './graphhoverline';
+import ValueAxis from './valueaxis';
+import TimeAxis from './timeaxis';
+
+const X_AXIS_HEIGHT = 20;
+const Y_AXIS_WIDTH = 35;
+const Y_AXIS_PADDING = 4;
+
+const getDatetimes = layers => (last(layers) ? last(layers).datapoints.map(d => d.data.datetime) : []);
+
+const getTimeScale = (containerWidth, datetimes, currentTime) => scaleTime()
+  .domain([first(datetimes), currentTime ? moment(currentTime).toDate() : last(datetimes)])
+  .range([0, containerWidth]);
+
+const getMaxTotalValue = layers => (last(layers) ? max(last(layers).datapoints.map(d => d[1])) : 0);
+
+const getValueScale = (containerHeight, maxTotalValue) => scaleLinear()
+  .domain([0, maxTotalValue * 1.1])
+  .range([containerHeight, Y_AXIS_PADDING]);
+
+const AreaGraph = React.memo(({
+  /*
+    `layers` should be an array of objects containing the following fields:
+      * `key` - unique ID of the layer
+      * `fill` - fill color of the layer
+      * `datapoints` - represents the graph datapoints + metadata for the layer in D3 stack() format
```

Comment:
Can we simplify even more and make sure the d3 stack step is done inside areaGraph?
This would make this component much more easy to use (and re-usable)

I'm thinking:
- data: array of datapoint objects. Each object represents a timestamp and must have a `datetime` key.
- layerAccessor: value or accessor function that returns a layer object from a data object.
- layerKeys: array of layer keys that should be present in layerAccessor. The order of this array defines the rendering order.
- layerFill: value or accessor function (e.g. key => some_color_mapping[d.key])
- layerHorizontalGradientFill: (e.g. datapoint => some_color_mapping[datapoint.cabonIntensity]) [do we need a gradient key here?]

Feel free to modify this based in order to optimize for performance (I don't remember what d3 stack requires)

Reply:
That's a good point, I'm glad you brought it up as it slipped my radar somehow under all these iterations of `AreaGraph` :)

I addressed it in https://github.com/tmrowco/electricitymap-contrib/pull/2171/commits/1a9c53ec2ba09a37c9280b59e36a244424de291c by:

- using `data` and `layerKeys` as you suggested
- using `layerFill` which maps layer keys to either a fixed color or a (datapoint -> color) map which then automatically creates a horizontal gradient
- not using `layerAccessor` and `layerHorizontalGradientFill` as they are obsolete with regards to the above

This change shouldn't have any effect on the performance. Please let me know what you think about the current props and whether you see any other potential blockers for merging the PR.